File: LC/128/128.java
/*
(100,4,200,1,3,2,5)
Find the starting point.
*/
class Solution {
    public int longestConsecutive(int[] nums) {
        if (nums.length == 0) return 0;
        
        int longest = 0;
        Set<Integer> set = new HashSet<>();
        for (int i = 0; i < nums.length; i++) set.add(nums[i]);
        
        for (int num: nums) {
            if (set.contains(num-1)) continue;
            
            int currNum = num;
            int currCount = 0;
            
            while (set.contains(currNum++)) currCount++;
            longest = Math.max(longest, currCount);
        }
        return longest;
    }   
}

File: LC/128/128.ts
/*
[100,4,200,1,3,2]

{
    100,
    4,
    200,
    1,
    3,
    2
    
}

99
*/
function longestConsecutive(nums: number[]): number {
    if (nums.length === 0) return 0;
    
    const set = nums.reduce((s, num) => {
        s.add(num);
        return s;
    }, new Set());
    
    let len = 0;
    for (let i = 0; i < nums.length; i++) {
        if (set.has(nums[i]-1)) continue;
        let nextNums = nums[i];
        while (set.has(nextNums)) nextNums++;
        len = Math.max(len, nextNums-nums[i]);
    }
    return len;
};
